Reformulating the grammar of rewstrategy with GRAMMAR EXTEND (#17832)
coq/coq: doc/sphinx/addendum/generalized-rewriting.rst

Usage
~~~~~

> .. tacn:: rewrite_strat {| @rewstrategy | {+; @rewstrategy } } {? in @ident }

Reviewer: ```suggestion
.. tacn:: rewrite_strat {+; @rewstrategy } {? in @ident }
```
right?

Author: It is produced automatically by docgram. It seems to correspond to a bug fix made by @jfehrle, but I'm not sure.

Reviewer 2: I'll take a look

Reviewer 2: When I download the branch, I see Jason's suggested change in the file.  Hugo, perhaps you did `doc_gram` rather than `doc_gram_rsts` or forget to include the rst file in your commit.

Author: I don't know. At least, the branch is right now synchronous.

.. tacn:: rewrite_strat @rewstrategy {? in @ident }
   :name: rewrite_strat

>    Rewrite using the specified :n:`@rewstrategy`s in the conclusion or in the hypothesis :n:`@ident`.

Reviewer: This should go back to how it was, right?
```suggestion
   Rewrite using :n:`@rewstrategy` in the conclusion or in the hypothesis :n:`@ident`.
```

on the programmable rewriting strategies with generic traversals by Visser et al.
:cite:`Luttik97specificationof` :cite:`Visser98`, which formed the core of
the Stratego transformation language :cite:`Visser01`. Rewriting strategies
are applied using the :tacn:`rewrite_strat` tactic`.

.. insertprodn rewstrategy rewstrategy_atom

.. prodn::
   rewstrategy ::= <- @one_term
   | subterms @rewstrategy
   | subterm @rewstrategy
   | innermost @rewstrategy
   | outermost @rewstrategy
   | bottomup @rewstrategy
   | topdown @rewstrategy
   | progress @rewstrategy
   | try @rewstrategy
   | any @rewstrategy
   | repeat @rewstrategy
   | old_hints @ident
   | hints @ident
   | terms {* @one_term }
   | eval @red_expr
   | fold @one_term
   | choice {+ @rewstrategy_atom }
   | @rewstrategy_atom
   rewstrategy_atom ::= @one_term
   | id
   | fail
   | refl

>    | ( {| @rewstrategy | {+; @rewstrategy } } )

Reviewer: This is redundant, right?  It should just be
```suggestion
   | ( {+; @rewstrategy } )
```

@jfehrle what do we need to make this happen?  (Or do you already know how to do this @herbelin ?)

Author: I don't know how to control how this is produced.

Reviewer: @jfehrle what do we need in doc/tools/docgram/common.edit_mlg to fix this (and the one above?)

on the programmable rewriting strategies with generic traversals by Visser et al.
:cite:`Luttik97specificationof` :cite:`Visser98`, which formed the core of
the Stratego transformation language :cite:`Visser01`. Rewriting strategies

> are applied using the :tacn:`rewrite_strat` tactic`.

Reviewer: ```suggestion
are applied using the :tacn:`rewrite_strat` tactic.
```

are applied using the :tacn:`rewrite_strat` tactic`.

.. insertprodn rewstrategy rewstrategy
.. insertprodn rewstrategy rewstrategy0

.. prodn::
   rewstrategy ::= @one_term

>    | <- @one_term

Reviewer 2: IIUC this production is entirely gone.  Doesn't that present a compatibility issue?

.. insertprodn rewstrategy rewstrategy

.. prodn::
   rewstrategy ::= @rewstrategy ; @rewstrategy_prefix_op

>    | @rewstrategy_prefix_op

Reviewer: The doc should define this, right?

Author: I haven't yet understood the semantics of `insertprod gram1 gram2`: it meant the range of entries from gram1 to gram2.

Hopefully, this is fixed. At least, none of `make doc/tools/docgram/fullGrammar doc/tools/docgram/updated_rsts doc/tools/docgram/orderedGrammar` complains. Double-check by @jfehrle would help tough.

:n:`@rewstrategy ; @rewstrategy`
   composition

> :n:`choice {+ @rewstrategy_atom }`

Reviewer: This is now
```suggestion
:n:`choice {+ @rewstrategy0 }`
```
right?

Reviewer 2: Indeed, this is incorrect, but this shouldn't be fixed by editing the rst.  I think Hugo didn't include the rst file changes in his last commit.  If you look in `orderedGrammar`, it has rewstrategy0 in this production.  Running `make doc_gram_rsts` will update the rst.  Then make sure it's included in the commit.

Author: This one is not generated and had to be done manually. Also fixed.

> :n:`choice {+ {+; @rewstrategy } }`

Reviewer: This seems wrong, we need parentheses now for sequencing with choice

Author: It is indeed wrong. It needed an update. Fixed.

:n:`@rewstrategy ; @rewstrategy`
   composition

> :n:`choice {+ @rewstrategy0 }`

Reviewer 2: Would you update the nonterminal names in the text (lines 840-958) so the references to `rewstrategy*` exactly match the grammar?  And maybe add some parens in the `choice` construct pseudocode on lines 920-926.

Author: Oh, I did not realize all that. Done.

I also tried to use `constr` in place of `glob` in the `mlg` and it seems to work.

> ``progress`` tests progress of the argument :n:`@rewstrategy` and

Reviewer 2: ```suggestion
``progress`` tests progress of the argument :n:`@rewstrategy1` and
```

fails if no progress was made, while ``try`` always succeeds, catching
failures. ``choice`` uses the first successful strategy in the list of

> :n:`@rewstrategy`s. One can iterate a strategy at least 1 time using

Reviewer 2: ```suggestion
:n:`@rewstrategy0`s. One can iterate a strategy at least 1 time using
```

   .. coqtop:: none

      Abort.

> .. note::
>    The name `fail` may refer both to the tactic :tac:`fail` and to
>    rewriting strategy `fail`. Similarly `;` may refer both to a
>    sequence of tactics and to a sequence of rewriting strategies. For
>    instance, `rewrite_strat s; fail` stands for `rewrite_strat (s; fail)`
>    by default. Use `(rewrite_strat s); fail` to mean a sequence of
>    tactics ending with the tactic :tac:`fail`.

Reviewer 2: **_Would you put this before the example (more consistent with rest of the doc)?_**

```suggestion
.. note::
   ';' is used to separate sequences of tactics as well as
   sequences of rewriting strategies.
   `rewrite_strat s; fail` is interpreted as `rewrite_strat (s; fail)`,
   in which `fail` is a rewriting strategy.
   Use `(rewrite_strat s); fail` to make `fail` a tactic.
   `rewrite_strat s; apply I` gives a syntax error (`apply` is not
   a valid rewrite strategy).
```

Author: > Would you put this before the example (more consistent with rest of the doc)?

Done.

For the suggestion, I don't know. Why not, but preferably not starting the sentence with a semicolon. So, maybe "The symbol ';' is used to separate sequences...".

Reviewer 2: Good point.  You could just say "Semicolons are used ..."
